libs/core/langchain_core/messages/ai.py

```diff
@@ -47,6 +48,87 @@ def __add__(self, other: Any) -> BaseMessageChunk:  # type: ignore
                     "Cannot concatenate AIMessageChunks with different example values."
                 )
 
+            content = merge_content(self.content, other.content)
+            additional_kwargs = merge_dicts(
+                self.additional_kwargs, other.additional_kwargs
+            )
+            response_metadata = merge_dicts(
+                self.response_metadata, other.response_metadata
+            )
+
+            if isinstance(other, AIToolCallsMessageChunk):
+                return AIToolCallsMessageChunk(
+                    example=self.example,
+                    content=content,
+                    additional_kwargs=additional_kwargs,
+                    response_metadata=response_metadata,
+                    tool_call_chunks=other.tool_call_chunks,
+                    id=self.id,
+                )
+
+            return self.__class__(
+                example=self.example,
+                content=content,
+                additional_kwargs=additional_kwargs,
+                response_metadata=response_metadata,
+                id=self.id,
+            )
+
+        return super().__add__(other)
+
+
+class ToolCall(Serializable):
+    name: str
+    args: dict
+    id: Optional[str] = None
+    index: Optional[int] = None
+
+
+class ToolCallChunk(Serializable):
+    name: Optional[str] = None
+    args: Optional[str] = None
+    id: Optional[str] = None
+    index: Optional[int] = None
+
+
+class AIToolCallsMessage(AIMessage):
+    tool_calls: Optional[List[ToolCall]] = None
+    type: Literal["tool_calls"] = "tool_calls"  # type: ignore[assignment] # noqa: E501
+
+
+class AIToolCallsMessageChunk(AIToolCallsMessage, AIMessageChunk):
+    # Ignoring mypy re-assignment here since we're overriding the value
+    # to make sure that the chunk variant can be discriminated from the
+    # non-chunk variant.
+    type: Literal["AIToolCallsMessageChunk"] = "AIToolCallsMessageChunk"  # type: ignore[assignment] # noqa: E501
+    tool_call_chunks: Optional[List[ToolCallChunk]] = None
```

Comment:
@baskaryan one issue with this: it's unclear what expected behavior for `message_chunk_to_message` should be without encoding specific parsing logic. Right now it generates an AIToolCallsMessage with tool_call_chunks. I'm not sure this is a problem though.

Reply:
could the parsing logic just be decoding json tool_call_chunks -> tool_calls?

Reply:
The args in the chunks are generally substrings of json. If we wanted to do partial parsing we'd need to move the  `parse_partial_json` function.